gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -535,7 +535,11 @@ public WorkUnit apply(@Nullable WorkUnit input) {
       } finally {
         try {
           TimingEvent jobCleanupTimer = this.eventSubmitter.getTimingEvent(TimingEvent.LauncherTimings.JOB_CLEANUP);
-          cleanupStagingData(jobState);
+          try {
+            cleanupStagingData(jobState);
+          } catch (Throwable t) {
+            LOG.error("Failed to clean up staging data", t);
```

Comment:
But this change will swallow the exception and any errors in `cleanupStagingData` will not fail the job and hence will go undetected (even if there is log message). 
Should we rather move the code we want to run (sending GTEs) to the finally block at line 600?

Reply:
Sure that's a good point, will change that

Reply:
Actually not sure if that's the correct place for it either, because if there is an error when notifying listeners, then the cleanup code to close the metrics reporting wouldn't run. Maybe it should be in the finally block at line 542

Reply:
Actually isn't this kind of the normal behaviour for a failed job? Any exception in work unit creation is caught at line 530 and not rethrown, does the job not failed for those cases either?

Reply:
Got it. Yes 542 is the right place.

Reply:
Okay, so the current behavior is do not fail the job (just mark the gobblin job as FAILED) when workunit creation fails, or anything in the main try block fails. However, according to the current behavior any exception in `cleanupStagingData` and `notifyListeners` are thrown (i.e. making job fail).
So, if we want to preserve the same behavior, we should move the code form L565-L599 to finally block at 542, right?
Why the current behavior treats `cleanupStagingData` failures differently, and whether we should change this behavior is a different question that I am not fully sure about.